File: src/sdk/javascript/examples/error-handling.ts
/**
 * 2api.ai JavaScript SDK - Error Handling Example
 *
 * Demonstrates proper error handling and retry logic.
 */

import {
  TwoAPI,
  TwoAPIError,
  AuthenticationError,
  RateLimitError,
  InvalidRequestError,
  ProviderError,
  TimeoutError,
  isRetryableError,
  RetryHandler,
} from 'twoapi';

async function main() {
  // ============================================================
  // Basic Error Handling
  // ============================================================
  console.log('=== Basic Error Handling ===\n');

  try {
    const client = new TwoAPI({ apiKey: 'invalid_key' });
    await client.chat('Hello');
  } catch (error) {
    if (error instanceof AuthenticationError) {
      console.log('Authentication failed:', error.message);
      console.log('Status code:', error.statusCode);
      console.log('Error code:', error.code);
    }
  }
  console.log();

  // ============================================================
  // Handling Different Error Types
  // ============================================================
  console.log('=== Error Type Detection ===\n');

  async function handleApiCall(client: TwoAPI, message: string) {
    try {
      return await client.chat(message);
    } catch (error) {
      if (error instanceof AuthenticationError) {
        // Don't retry - fix the API key
        console.error('Auth Error: Check your API key');
        throw error;
      }

      if (error instanceof RateLimitError) {
        // Wait and retry
        console.log(`Rate limited. Retry after ${error.retryAfter} seconds`);
        // In a real app, you might wait and retry here
        throw error;
      }

      if (error instanceof InvalidRequestError) {
        // Don't retry - fix the request
        console.error('Invalid Request:', error.message);
        throw error;
      }

      if (error instanceof ProviderError) {
        // Provider is down - might retry or use fallback
        console.error(`Provider ${error.provider} error:`, error.message);
        throw error;
      }

      if (error instanceof TimeoutError) {
        // Request timed out - might retry
        console.error('Request timed out');
        throw error;
      }

      if (error instanceof TwoAPIError) {
        // Generic API error
        console.error('API Error:', error.message);
        throw error;
      }

      // Unknown error
      throw error;
    }
  }

  // ============================================================
  // Using isRetryableError
  // ============================================================
  console.log('=== Checking Retryability ===\n');

  const errors = [
    new AuthenticationError('Invalid key'),
    new RateLimitError('Too many requests', 429),
    new InvalidRequestError('Bad params', 400),
    new ProviderError('OpenAI down', 503),
    new TimeoutError('Timed out'),
  ];

  for (const error of errors) {
    console.log(
      `${error.constructor.name}: retryable = ${isRetryableError(error)}`
    );
  }
  console.log();

  // ============================================================
  // Custom Retry Configuration
  // ============================================================
  console.log('=== Custom Retry Configuration ===\n');

  const client = new TwoAPI({
    apiKey: process.env.TWOAPI_API_KEY,
    maxRetries: 5, // Try up to 5 times
    onRetry: (attempt, error, delay) => {
      console.log(`Retry attempt ${attempt + 1} after ${delay}ms: ${error.message}`);
    },
  });

  // The client will automatically retry retryable errors
  try {
    const response = await client.chat('Hello!');
    console.log('Success:', response.content);
  } catch (error) {
    console.error('All retries failed:', (error as Error).message);
  }
  console.log();

  // ============================================================
  // Manual Retry Handler
  // ============================================================
  console.log('=== Manual Retry Handler ===\n');

  const retryHandler = new RetryHandler({
    maxRetries: 3,
    initialDelay: 1000, // 1 second
    maxDelay: 30000, // 30 seconds max
    exponentialBase: 2,
    onRetry: (attempt, error, delay) => {
      console.log(`Manual retry ${attempt + 1}, waiting ${delay}ms`);
    },
  });

  async function makeReliableRequest() {
    return retryHandler.execute(async () => {
      // Your API call here
      const client = new TwoAPI();
      return client.chat('Hello!');
    });
  }

  try {
    const response = await makeReliableRequest();
    console.log('Reliable request succeeded:', response.content);
  } catch (error) {
    console.error('Reliable request failed:', (error as Error).message);
  }
  console.log();

  // ============================================================
  // Error Details
  // ============================================================
  console.log('=== Error Details ===\n');

  try {
    const client = new TwoAPI({ apiKey: 'bad_key' });
    await client.chat('Hello');
  } catch (error) {
    if (error instanceof TwoAPIError) {
      console.log('Error Details:');
      console.log('  Message:', error.message);
      console.log('  Status Code:', error.statusCode);
      console.log('  Error Code:', error.code);
      console.log('  Is Retryable:', isRetryableError(error));
    }
  }
}

main().catch(console.error);

File: src/sdk/python/examples/error_handling.py
"""
2api.ai Python SDK - Error Handling Example

Demonstrates proper error handling and retry logic.
"""

import os
from twoapi import TwoAPI
from twoapi.errors import (
    TwoAPIError,
    AuthenticationError,
    RateLimitError,
    InvalidRequestError,
    ProviderError,
    TimeoutError,
    is_retryable_error,
)
from twoapi.retry import RetryHandler, calculate_backoff

def main():
    # ============================================================
    # Basic Error Handling
    # ============================================================
    print("=== Basic Error Handling ===\n")

    try:
        client = TwoAPI(api_key="invalid_key")
        client.chat("Hello")
    except AuthenticationError as e:
        print(f"Authentication failed: {e.message}")
        print(f"Status code: {e.status_code}")
        print(f"Error code: {e.code}")
    print()

    # ============================================================
    # Handling Different Error Types
    # ============================================================
    print("=== Error Type Detection ===\n")

    def handle_api_call(client: TwoAPI, message: str):
        try:
            return client.chat(message)
        except AuthenticationError as e:
            # Don't retry - fix the API key
            print(f"Auth Error: Check your API key - {e.message}")
            raise

        except RateLimitError as e:
            # Wait and retry
            print(f"Rate limited. Retry after {e.retry_after} seconds")
            # In a real app, you might wait and retry here
            raise

        except InvalidRequestError as e:
            # Don't retry - fix the request
            print(f"Invalid Request: {e.message}")
            raise

        except ProviderError as e:
            # Provider is down - might retry or use fallback
            print(f"Provider {e.provider} error: {e.message}")
            raise

        except TimeoutError as e:
            # Request timed out - might retry
            print(f"Request timed out: {e.message}")
            raise

        except TwoAPIError as e:
            # Generic API error
            print(f"API Error: {e.message}")
            raise

    # ============================================================
    # Using is_retryable_error
    # ============================================================
    print("=== Checking Retryability ===\n")

    errors = [
        AuthenticationError("Invalid key"),
        RateLimitError("Too many requests", status_code=429),
        InvalidRequestError("Bad params", status_code=400),
        ProviderError("OpenAI down", status_code=503),
        TimeoutError("Timed out"),
    ]

    for error in errors:
        print(f"{error.__class__.__name__}: retryable = {is_retryable_error(error)}")
    print()

    # ============================================================
    # Custom Retry Configuration
    # ============================================================
    print("=== Custom Retry Configuration ===\n")

    def on_retry(attempt: int, error: Exception, delay: float):
        print(f"Retry attempt {attempt + 1} after {delay:.2f}s: {error}")

    client = TwoAPI(
        api_key=os.getenv("TWOAPI_API_KEY"),
        max_retries=5,  # Try up to 5 times
        on_retry=on_retry,
    )

    # The client will automatically retry retryable errors
    try:
        response = client.chat("Hello!")
        print(f"Success: {response.content}")
    except TwoAPIError as e:
        print(f"All retries failed: {e.message}")
    print()

    # ============================================================
    # Manual Retry Handler
    # ============================================================
    print("=== Manual Retry Handler ===\n")

    def manual_on_retry(attempt: int, error: Exception, delay: float):
        print(f"Manual retry {attempt + 1}, waiting {delay:.2f}s")

    retry_handler = RetryHandler(
        max_retries=3,
        initial_delay=1.0,  # 1 second
        max_delay=30.0,  # 30 seconds max
        exponential_base=2.0,
        on_retry=manual_on_retry,
    )

    def make_reliable_request():
        def _request():
            client = TwoAPI()
            return client.chat("Hello!")
        return retry_handler.execute(_request)

    try:
        response = make_reliable_request()
        print(f"Reliable request succeeded: {response.content}")
    except TwoAPIError as e:
        print(f"Reliable request failed: {e.message}")
    print()

    # ============================================================
    # Backoff Calculation
    # ============================================================
    print("=== Backoff Calculation ===\n")

    print("Exponential backoff delays:")
    for attempt in range(5):
        delay = calculate_backoff(attempt, initial_delay=1.0, jitter=False)
        delay_jitter = calculate_backoff(attempt, initial_delay=1.0, jitter=True)
        print(f"  Attempt {attempt}: {delay:.2f}s (with jitter: {delay_jitter:.2f}s)")
    print()

    # ============================================================
    # Error Details
    # ============================================================
    print("=== Error Details ===\n")

    try:
        client = TwoAPI(api_key="bad_key")
        client.chat("Hello")
    except TwoAPIError as e:
        print("Error Details:")
        print(f"  Message: {e.message}")
        print(f"  Status Code: {e.status_code}")
        print(f"  Error Code: {e.code}")
        print(f"  Is Retryable: {is_retryable_error(e)}")
        print(f"  Request ID: {e.request_id}")
        print(f"  Details: {e.details}")


if __name__ == "__main__":
    main()
